### PR TITLE
Reincarnation

### PR DESCRIPTION
**Pull request by Ingvord**
As many users request TangoSourceDistribution, see, for instance, [cppTango issue #450](https://github.com/tango-controls/cppTango/issues/450)

We need to update this repo to take into account migration to github, cmakenzation for cpp projects and mavenization of Java

1. [x] provide a script that will fetch required tagged source codes and put them into a predefined folder structure
2. [x] later script must assemble source distribution and deploy it to GitHub releses
3. [x] update auto tools (or replace them) to build cpp using cmake
4. [x] update auto tools (or replace them)  to fetch Java from maven
5. [x] setup continuous integration

### CODE REVIEW

**Comment by Ingvord** [comment]
OK, we now have source distribution where at least `./configure` works -- clone, checkout and run ant.

I have not tested installation yet as I plan to migrate to cmake anyway. Those who want to play with `./configure` may take dc6eab6 as a basis.

A few observations:

1) there is no latest pogo release in maven (9.6.14) so distribution takes 9.6.12
2) pogo/templates seem to be obsolete
3) TangoAccessControl has not been released after migration to GitHub -> _master_ is used
4) Some CMake files (especially ones generated by pogo) are environment depended:

```
# MAKE_ENV is the path to find common environment to buil project
#
set(MAKE_ENV /segfs/tango/cppserver/env)
```

Making them basically useless

@bourtemb , your comments are welcome!

**Comment by Ingvord** [comment]
I have add CMake files to see if we can painlessly replace autotools with CMake. See 5d066c8 and ca5a30b

Here are some thoughts:

1) currently we have the following dependency chain for c++ projects: idl < lib < cppserver/utils. For simplicity, I would suggest to split installation into three steps: install idl; install lib; install the rest. Dunno if this can be done in CMake easily, for instance, now to build e.g. tango_admin (provided that tango was pre-installed)

2) Some C++ components do not have CMake files others have CMake files that are environment depended. So additional work is required here

Otherwise it seems promising. We will use CMake to compile and install all the required components which is perfect for GNU based OSes. Could also work out for Windows.

File: assets/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8.12)
 project(TangoSourceDistribution)
 
+set(IDL_BASE ${PROJECT_SOURCE_DIR}/lib/idl)
```
Comment (bourtemb) [inline assets/CMakeLists.txt]: Why don't we install the IDL somewhere and use this installation directory as IDL_BASE?
Then we don't need to change IDL_BASE definition in cppTango...

Comment (Ingvord) [reply assets/CMakeLists.txt]: Sure, I am just playing with it. Basically we need to install not only idl, but tango, before we can try to  compile other Tango components

**Comment by bourtemb** [comment]
@ingvord, could you please add a README file which explains how to use this new version of the distribution?
I didn't manage to execute successfully the configure script yet... but it might be because I missed some mandatory preliminary steps.

Please note that if we keep the configure script, it would be good to change its permissions in git to make it executable.

@Pascal-Verdier, could you please suggest what version of POGO we should take to prepare the next Tango distribution (see @ingvord above comment: https://github.com/tango-controls/TangoSourceDistribution/pull/8#issuecomment-424758391)?
I think it would be good to tag a release for TangoAccessControl on Github too.

**Comment by Ingvord** [comment]
@bourtemb , I tagged the code base where configure works - `git checkout autotools`. Later commits won't work because repository structure has changed in cppTango and some others.

**Comment by Ingvord** [comment]
configure script fails to recognize mariadb/mysql version on debian 9.5.0 (stretch):

```
checking for mysql... /usr/bin/mysql
checking for mysql release (at least version 5.5)... no
configure: WARNING: Not supported mysql release. Should be 5.5.x or above. Please update !
```

while mysql-server is 5.5.9999:

```
ingvord@vb-stretch:~/Downloads$ sudo apt-get install mysql-server 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  default-mysql-server mariadb-server-10.1
Suggested packages:
  mailx mariadb-test netcat-openbsd tinyca
The following NEW packages will be installed:
  default-mysql-server mariadb-server-10.1 mysql-server
0 upgraded, 3 newly installed, 0 to remove and 1 not upgraded.
Need to get 4,754 B/5,338 kB of archives.
After this operation, 78.5 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 http://deb.debian.org/debian stable/main amd64 default-mysql-server all 1.0.2 [3,048 B]
Get:2 http://deb.debian.org/debian stable/main amd64 mysql-server amd64 5.5.9999+default [1,706 B]
Fetched 4,754 B in 0s (80.7 kB/s)
Preconfiguring packages ...
Selecting previously unselected package mariadb-server-10.1.
(Reading database ... 170592 files and directories currently installed.)
Preparing to unpack .../mariadb-server-10.1_10.1.26-0+deb9u1_amd64.deb ...
/var/lib/mysql: found previous version 10.1
Unpacking mariadb-server-10.1 (10.1.26-0+deb9u1) ...
Selecting previously unselected package default-mysql-server.
Preparing to unpack .../default-mysql-server_1.0.2_all.deb ...
Unpacking default-mysql-server (1.0.2) ...
Selecting previously unselected package mysql-server.
Preparing to unpack .../mysql-server_5.5.9999+default_amd64.deb ...
Unpacking mysql-server (5.5.9999+default) ...
Processing triggers for systemd (232-25+deb9u4) ...
Processing triggers for man-db (2.7.6.1-2) ...
Setting up mariadb-server-10.1 (10.1.26-0+deb9u1) ...
Setting up default-mysql-server (1.0.2) ...
Setting up mysql-server (5.5.9999+default) ...
```

and

```
ingvord@vb-stretch:~/Downloads$ mysql --version
mysql  Ver 15.1 Distrib 10.1.26-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2
```

**Comment by Ingvord** [comment]
OK, I have managed to build everything using old configure script. The problem now is with DB when installing:

```
make[2]: Entering directory '/home/ingvord/Downloads/tango-9.3.2/cppserver/database'
/bin/bash ../../cppserver/database/create_db.sh < ../../cppserver/database/create_db.sql
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)
Makefile:781: recipe for target 'all-local' failed
make[2]: *** [all-local] Error 1
make[2]: Leaving directory '/home/ingvord/Downloads/tango-9.3.2/cppserver/database'
Makefile:440: recipe for target 'install-recursive' failed
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory '/home/ingvord/Downloads/tango-9.3.2/cppserver'
Makefile:530: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1
ingvord@vb-jessie:~/Downloads/tango-9.3.2$
```

I will reset this branch to discard CMake related commits and basically that is it...

I would also clean up other repos - to remove all make related files, so independently libraries must be built using CMake and not to have a conflict with Makefiles generated by configure in this repo.

A few more observations:

- latest JTango release fat jar file is missing. Reported [here](https://github.com/tango-controls/JTango/issues/61)
- CMake build is required prior cppTango move target
- zmq.hpp is in lib/client - do we still need this?
- Travis fails to configure distribution due to missing libzmq even though libzmq3-dev package is installed:

```
checking for LIBZMQ... no
configure: error: Missing ZMQ library - Use --with-zmq configure option or add directory containing libzmq.pc to PKG_CONFIG_PATH environment variable
The command "./configure" exited with 1.
```

**Comment by Ingvord** [comment]
OK, I have manged to build everything and install on a fresh Debian 8.11.0 (jessie) VB (using --prefix). 

So this is now ready for merging. But first I would merge #2, #3, #4 and #5 then rebase or merge this PR with master, so that not lose these PRs (#2, #3, #4 and #5).

There might be (and there are) some issues, like `./configure` does not work on Debian 9. These can be fixed via dedicated PRs.

**Comment by Pascal-Verdier** [comment]
Hi
A new Pogo release has been send to BinTray (9.6.16).
About the obsolescence of Pogo templates , the c++ level (11/14/17/...) we want must be discussed in kernel meeting.
Pascal

**Comment by Pascal-Verdier** [comment]
About 
set(MAKE_ENV /segfs/tango/cppserver/env)
It is not a problem with Pogo, but with the installation:
Pogo gets information from a file named: Pogo.site_properties
In this file the property  **org.tango.pogo.makefile_home**  must define the path to find common environment to build project.

**Comment by Ingvord** [comment]
@Pascal-Verdier 

> A new Pogo release has been send to BinTray (9.6.16).

Confirm. It is in bintray now. Thanks!

> It is not a problem with Pogo, but with the installation:

IMHO, CMake files must be environment independent, that means no global properties or any other magic properties unless it is common infrastructure, like pkg-config. And especially ones that reside in public repos. Users must be able to clone let say TangoAccessControl and run cmake to build it.

> About the obsolescence of Pogo templates , the c++ level (11/14/17/...) we want must be discussed in kernel meeting.

I also could not find these in Pogo repo, is this repo the only place where they reside? If so I would suggest to move them to [Pogo](https://github.com/tango-controls/pogo) as they definitely belong to Pogo development lifecycle. And this one must be a dumm thing that just fetches resources from other repos.

File: assets/configure [hunk 2]
```diff
@@ -3301,7 +3301,7 @@ fi
 
 # Define the identity of the package.
  PACKAGE='tango'
- VERSION='9.2.5'
+ VERSION='@VERSION@'
```
Comment (bourtemb) [inline assets/configure]: As I already wrote in another comment, we should not modify the configure script.
We should rather change configure.ac and use the boostrap script to generate a fresh configure script with the autotools.

**Comment by Ingvord** [comment]
@bourtemb , ant target `bootstrap-distribution` has been added; Makefile.in and configure files have been removed from repo.  I have tested on Debian 8 the resulting distribution i.e. tango-autotools.tar.gz (configuring, making and installation) - looks good.

**Comment by bourtemb** [comment]
Hi,

I tried on a Debian Stretch VM with the new boostrap-distribution ant target.
I got a tango-dev.tar.gz generated file (I tried as well with tango-autotools.tar.gz fron github releases page).
I extracted it in another directory and I did:
```
cd tango-9.3.2
chmod +x configure
mkdir build
cd build
../configure
```

The configure step went well but I got the following error when trying to compile:

```
reynald@debian9-vm:~/test_distrib/tango-9.3.2/build$ make
make  all-recursive
make[1]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build'
Making all in lib
make[2]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build/lib'
Making all in cpp
make[3]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp'
Making all in log4tango
make[4]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango'
Making all in config
make[5]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango/config'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango/config'
Making all in src
make[5]: Entering directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango/src'
/bin/bash ../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../../../../../lib/cpp/log4tango/src -I.. -I../../../../../lib/cpp/log4tango/include    -g -O2 -Wall -Wno-unused -pedantic -D_REENTRANT -MT Appender.lo -MD -MP -MF .deps/Appender.Tpo -c -o Appender.lo ../../../../../lib/cpp/log4tango/src/Appender.cpp
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../../../../../lib/cpp/log4tango/src -I.. -I../../../../../lib/cpp/log4tango/include -g -O2 -Wall -Wno-unused -pedantic -D_REENTRANT -MT Appender.lo -MD -MP -MF .deps/Appender.Tpo -c ../../../../../lib/cpp/log4tango/src/Appender.cpp  -fPIC -DPIC -o .libs/Appender.o
In file included from ../../../../../lib/cpp/log4tango/src/PortabilityImpl.hh:31:0,
                 from ../../../../../lib/cpp/log4tango/src/Appender.cpp:29:
../../../../../lib/cpp/log4tango/include/log4tango/Portability.hh:34:31: fatal error: log4tango/config.h: No such file or directory
 # include <log4tango/config.h>
                               ^
compilation terminated.
Makefile:587: recipe for target 'Appender.lo' failed
make[5]: *** [Appender.lo] Error 1
make[5]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango/src'
Makefile:504: recipe for target 'all-recursive' failed
make[4]: *** [all-recursive] Error 1
make[4]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp/log4tango'
Makefile:494: recipe for target 'all-recursive' failed
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build/lib/cpp'
Makefile:441: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build/lib'
Makefile:530: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/reynald/test_distrib/tango-9.3.2/build'
Makefile:460: recipe for target 'all' failed

```

This error seems to be easy to reproduce.
Simply create a build directory and execute configure from this directory.
We should be able to do that. It is especially useful when you want to build for different platforms from the same source distribution.

**Comment by Ingvord** [comment]
@bourtemb , I can reproduce this on Jessie as well, looks like `-I../include` direcrtive is missing

**Pull request merged**